ls1intum/Artemis
# [Feature] Introducing the Editor Role

### Checklist
- [x] Server: I adapted existing integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I adapted `@PreAuthorize` and check the course groups for all related REST Calls (security)
- [x] Client: I adapted existing integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I edited `authorities` for all routes and checked the course groups for displaying navigation elements (links, buttons)
- [x] Client: I added multiple screenshots/screencasts of (some of) my UI changes

### Motivation and Context


Due to the increasing number of students, it is necessary for tutors to participate in the creation and maintenance of exercises. These tutors are so-called **Editors**.
The current role system does not offer a dedicated _Editor_ role, meaning that editors are granted _Instructor_ rights to be able to create and edit exercises. This is problematic because _Editors_ can access more functionalities than necessary. In order to avoid sources of error, such as the deletion of exercises or changes to the grading the permissions of the _Editor_ should be limited.

### Description

The new role of an _**Editor**_ is introduced. (_Instructor_ > _Editor_ > _TeachingAssistant/Tutor_)
The _Editor_ is able to create and update exercises but can not delete those or affect the grading.

### Steps for Testing

Functionalities an _Editor_ can access
- test for normal course management and exams
  - create exercise (text, modeling, quiz, programming, file upload)
  - edit exercises (e.g. adjusting the problem statement, adjusting settings of the exercise, changing release / due dates)
- check access rights to pages
  - editors can access at least all pages that tutors can access
  - tutor and instructor access rights did not change

Known issue:
_The access rights to programming exercises (the repositories and projects) are not retroactivly adapted when changing group names / adding groups to a course. This issue https://github.com/ls1intum/Artemis/issues/3327 is out of scope for this PR and already existed before.
For newly created exercises the access rights are set correctly._

### Test Coverage




### Screenshots



#### Example: Course Management Exercises
_Editors_ can access at least all the functionality that is provided to _tutors_, can edit exercises but are not able to delete them and have limited access to student information.

_Tutor_
![image](https://user-images.githubusercontent.com/63976129/115992043-d545d600-a5cb-11eb-89ab-28c7192d78bb.png)

_Editor_
![image](https://user-images.githubusercontent.com/63976129/115991938-4b960880-a5cb-11eb-9955-00fb22968737.png)

_Instructor_
![image](https://user-images.githubusercontent.com/63976129/115991970-6c5e5e00-a5cb-11eb-93dd-f6f9c989a385.png)

### Coverage
BuildFeature: 63
BuildDevelop: 2439

Server Tests: 88% (+0)
Client Tests: 75,69% (-0,01)
